### ISSUE TITLE
TRS spec should provide guidance for TRS consumers and trs:order

### ISSUE THREAD

**Issue by DavidJHoney** [oslc-op/oslc-specs#20]
Currently the spec says that TRS producers must ensure that the trs:order values must reflect the sequence in which events became visible in the change log. In practice, this is quite difficult. Many applications will want to create the events describing new, modified, and deleted tracked resources in the same transaction as the new, modified, and deleted tracked resources. Currently, all of the implementations I have seen allocate the next trs:order within that transaction. The problem is that if the transactions start in some order, their commits may complete in a different order. For example, consider the following sequence:

Time|Thread 1|Thread 2
----|--------|--------
T1|Start transaction TX1|
T2|Modified tracked resource R1|
T3|Allocate trs:order 1|
T4||Start transaction TX2
T5||Modify tracked resource R2
T6|Start commit of TX1|
T7||Allocate trs:order 2
T8||Start commit of TX2
T9||TX2 commit completed
T10|TX1 commit completed|

The problem is that at time T9, the TRS change log contains the event with trs:order 2, and at the later time T10, a lower trs:order 1 is present. Applications have no control over the order in which transactions _complete_.

Strict conformance to the TRS specification imposes significant implementation burden and cost. It would require that TRS events created in a transaction are not visible in the change log even after TX commit, and that a separate single thread then processes these as-yet-unseen events to allocate trs:order, even in a clustered environment. Such a single thread could become a bottleneck because it cannot process the events as fast as multiple threads are creating new events.

I think the spec needs to recommend some leniency for consumers of the TRS, recommending that they should not assume that all events become visible in trs:order within the small time windows that can arise from the reasons stated above. This remains a potential source of events being skipped or mishandled. Nick and I discussed this, and it was agreed that the specification should provider guidance for such leniency in consuming the change log.

**Comment by berezovskyi** [comment]
Okay, so we implemented TRS over MQTT we decided not to trust any of that and inline the whole resource so that the consumer can obtain consistent state of the system at the time of the CE generation. Without versioning or inlining, no amount of spec changes will get good consistency. Having said that, the (generated) order of the CEs should not matter that much.

What should matter is causal consistency (ie you don't order a Deletion CE for the same resource before its Creation event). However, how a clustered environment does that is beyond the scope of the spec.

Now, I think there might be some misunderstanding regarding the purpose of the TRS Change Logs if we are suggesting for the TRS Clients to ignore errors while retrieving resources. TRS is used to announce changes to resources. Thus, I see no problem if the TRS CL lags 200ms behind the REST I/F or so.

Now, you talk about the bottleneck. I am afraid it's hard for me to see how can you get one. An implementation of a distributed (!) atomic long counter allows you to do upwards of 100k reads/s ([link](https://hazelcast.com/blog/just-made-hazelcast-3-7-30-faster/)). Yes, a hundred thousand reads a second. Which would BTW give you an ability to impose a strict sequential ordering on the trs:order rather just the one of monotonic growth.

> Just to be clear: here I would suggest to run transactions w/o generating trs:order IDs. Once the TX commits, we put the successful TX CEs into the in-mem queue (one per node in a cluster) and drain the queue, using the distributed atomic long to assign globally strictly sequential IDs.

Sorry for discussing too many things at a time, I just wanted to answer each of your points.

Bottom line: I suggest not to introduce such leniency and maybe even make trs:order requirements more strict.

**Comment by berezovskyi** [comment]
BTW, I shall admit that I also hit the same problem when I realised that my TRS CEs were sent over MQTT even before the SPARQL UPDATE has succeeded and I had to rewrite a bit of the code to do a blocking request and to issue IDs later in the process as I described.

**Comment by berezovskyi** [comment]
Redis INCR command also [allows](https://redis.io/topics/benchmarks) around 100k ops/s.

**Comment by DavidJHoney** [comment]
Be careful with including resource content in the payload of MQTT messages:
1) MQTT makes no guarantees about the order in which a subscribed client will get notifications.
2) Embedding such content means customers would need to use secure protocols, and many MQTT implementations recommend not doing so because it has a signifcant performance impact on the broker.

**Comment by berezovskyi** [comment]
Thanks! I do reordering based on trs:order on the client side but so far has not got any problems. Apparently, MQTT [DOES make ordering guarantees](https://stackoverflow.com/questions/30955110/is-message-order-preserved-in-mqtt-messages) for QoS 0 and QoS 2 (we use QoS 2) but not QoS 1 because older messages can be deliver multiple times.

MQTT uses TLS just as HTTP would, so performance impact is either poor implementation or FUD (I would presume). UPD: just looked at [this](https://www.hivemq.com/blog/how-does-tls-affect-mqtt-performance/) post and TLS has impact on the CPU from the handshake. Just as HTTP would. But I think the TLS overhead is even lower for MQTT than for HTTPS because the connections remain open for longer periods of time and the initial perfomance cost can be ammortised better. And Lyo Client that ignores TLS cert validity also has a lot of room to grow :)

**Comment by DavidJHoney** [comment]
The issue about "bottleneck" is that to guarantee ascending trs:order values with respect to time, an implementation either has to do so in a single thread or use synchronization (including across cluster nodes). Just using AtomicLong is insufficient. The critical section includes fetching the event from the repository, allocation the trs:order, and saving the updated event in the repository. That critical section needs to atomic and non-current.

There are two choices here:
1) Perform the trs:order allocation on the requesting thread _after_ the TX commit. The problem here is that this needs to be synchronized, including across cluster nodes. That cluster-wide lock becomes a source of contention where multiple threads have to wait for the lock to become available. In turn, this means that operations the user initiated have performance that doesn't scale well.

2) Perform the allocation of trs:order outside of the original thread. Users don't see performance of operations being impated. But now you have one, or more than one, thread that runs asynchronously trying to allocatetrs:orders, with the same issue for lock contention. This would mean that if there was a high rate of tracked resource changes, there might be a backlog of TRS events waiting to be alloated trs:order values, and hence a TRS consumer might see a delay in changes that might increase until the rate of tracked resource changes drops and the background tasks can catch up with the backlog.

**Comment by DavidJHoney** [comment]
Re: MQTT ordering. This looks like a new feature in MQTT 3.1 versus 2.x.
I recall seeing discussions of QOS 2 that it imposed an overhead, and hence from a performance viewpoint, QOS 1 might be preferable. Jazz applications use MQTT for cache invalidation, and use QOS 1 for performance, and do not include resource data in the payload so that secure protocols are not required, and order of delivery is then unimportant.

**Comment by berezovskyi** [comment]
I actually mean both of those options to be used.

1. First, you start a thread as you describe under the option no. 2.
2. That thread will actually handle CEs after the resource persistence TX has been committed. This is were we actually do the actions you describe in step no. 1.

A correction for the atomic long: I did not suggest using Java AtomicLong but instead a distributed atomic long implementation. It is synchronised across the cluster and the tests show it to be able to allow 100k reads/s. I would be OK with that performance for the majority of the cases. Higher-performance cases might require a distributed log like Kafka to push beyond 100k writes/s.

There is an even better data structure based on the conflict-free replicated data type https://docs.hazelcast.org/docs/latest-dev/manual/html-single/index.html#pn-counter

**Comment by ndjc** [comment]
The TRS 3.0 draft maintains the ordering requirement.
